src/elf_parser_glue.rs

```diff
@@ -0,0 +1,546 @@
+        consts::{SHF_ALLOC, SHF_WRITE, SHT_NOBITS, STT_FUNC},
+        types::{
+            Elf64Addr, Elf64Ehdr, Elf64Off, Elf64Phdr, Elf64Rel, Elf64Shdr, Elf64Sym, Elf64Word,
+            Elf64Xword, ElfIdent,
+        },
+        Elf64, ElfParserError,
+    },
+    error::{EbpfError, UserDefinedError},
+};
+
+/// The common trait implemented by LegacyParser and NewParser.
+///
+/// This is an internal interface used to isolate the ELF parsing bits and to be
+/// able to plug the old goblin parser or the new parser depending on config.
+///
+/// The interface is pretty straightforward. The associated types are the types
+/// used to represent ELF data. Some return values are `Cow<T>` since goblin
+/// returns some data by value, while the new parser always borrows from the
+/// underlying file slice.
+pub trait ElfParser<'a>: Sized {
+    /// Program header type.
+    type ProgramHeader: ElfProgramHeader + 'a;
+    /// Iterator of program headers.
+    type ProgramHeaders: Iterator<Item = &'a Self::ProgramHeader>;
+
+    /// Section header type.
+    type SectionHeader: ElfSectionHeader + 'a;
+    /// Iterator of section headers
+    type SectionHeaders: Iterator<Item = &'a Self::SectionHeader>;
+
+    /// Symbol type.
+    type Symbol: ElfSymbol + 'a;
+    /// Iterator of symbols.
+    type Symbols: Iterator<Item = Cow<'a, Self::Symbol>>;
+
+    /// Relocation type.
+    type Relocation: ElfRelocation + 'a;
+    /// Iterator of relocations.
+    type Relocations: Iterator<Item = Cow<'a, Self::Relocation>>;
+
+    /// Parses the ELF data included in the buffer.
+    fn parse(data: &'a [u8]) -> Result<Self, ElfError>;
+
+    /// Returns the file header.
+    fn header(&self) -> &Elf64Ehdr;
+
+    /// Returns the program headers.
+    fn program_headers(&'a self) -> Self::ProgramHeaders;
+
+    /// Returns the section headers.
+    fn section_headers(&'a self) -> Self::SectionHeaders;
+
+    /// Returns the section with the given `name`.
+    fn section(&self, name: &str) -> Result<Self::SectionHeader, ElfError>;
+
+    /// Returns the section name at the given `sh_name` offset.
+    fn section_name(&self, sh_name: Elf64Word) -> Option<&str>;
+
+    /// Returns the symbols included in the symbol table.
+    fn symbols(&'a self) -> Self::Symbols;
+
+    /// Returns the `index`th symbol in the symbol table.
+    fn symbol(&self, index: usize) -> Option<Self::Symbol>;
+
+    /// Returns the symbol name at the given `st_name` offset.
+    fn symbol_name(&self, st_name: Elf64Word) -> Option<&str>;
+
+    /// Returns the symbols included in the dynamic symbol table.
+    fn dynamic_symbol(&self, index: Elf64Word) -> Option<Self::Symbol>;
+
+    /// Returns the dynamic symbol name at the given `st_name` offset.
+    fn dynamic_symbol_name(&self, st_name: Elf64Word) -> Option<&str>;
+
+    /// Returns the dynamic relocations.
+    fn dynamic_relocations(&'a self) -> Self::Relocations;
+}
+
+/// ELF program header.
+#[allow(missing_docs)]
```

Comment:
Why do we need these?

Reply:
Oh, we don't anymore! I added it at some point to shut up rust analyzer, but then I did add the docs